File: benchmarks/workloads/buffers/buffers.go
// Buffers: chunked-I/O shaped churn of 8 to 64 KiB blocks, 32 live at once.
// Timed region: everything. Checksum: wrapping sum of the words written.
package main

import (
	"fmt"
	"time"
)

const (
	iterations = 2000000
	ring       = 32
)

func main() {
	start := time.Now()
	slots := make([][]uint64, ring)
	var checksum uint64
	for i := 0; i < iterations; i++ {
		words := 1024 * (1 + (i % 8))
		block := make([]uint64, 0, words)
		for j := 0; j < words/64; j++ {
			value := uint64(i)*2654435761 + uint64(j)
			block = append(block, value)
			checksum += value
		}
		slots[i%ring] = block
	}
	elapsedMs := float64(time.Since(start).Nanoseconds()) / 1e6
	fmt.Printf("iterations %d ring %d\n", iterations, ring)
	fmt.Printf("elapsed_ms %.3f\n", elapsedMs)
	fmt.Printf("checksum %d\n", checksum)
}

File: benchmarks/workloads/buffers/buffers.rs
// Buffers: chunked-I/O shaped churn of 8 to 64 KiB blocks, 32 live at once.
// Timed region: everything. Checksum: wrapping sum of the words written.
const ITERATIONS: usize = 2_000_000;
const RING: usize = 32;

fn main() {
    let start = std::time::Instant::now();
    let mut ring: Vec<Vec<u64>> = (0..RING).map(|_| Vec::new()).collect();
    let mut checksum: u64 = 0;
    for i in 0..ITERATIONS {
        let words = 1024 * (1 + (i % 8));
        let mut block: Vec<u64> = Vec::with_capacity(words);
        for j in 0..(words / 64) {
            let value = (i as u64).wrapping_mul(2654435761).wrapping_add(j as u64);
            block.push(value);
            checksum = checksum.wrapping_add(value);
        }
        ring[i % RING] = block;
    }
    let elapsed_ms = start.elapsed().as_secs_f64() * 1000.0;
    println!("iterations {} ring {}", ITERATIONS, RING);
    println!("elapsed_ms {:.3}", elapsed_ms);
    println!("checksum {}", checksum);
}
